README.md
`delay` is a shell command introducing a constant delay between its standard
input and its standard output.

## Usage

    delay [-b <dtbufsize>] <delay>

`dtbufsize` is the buffer size storing the data until it has been written to
stdout, in bytes.

The following modifiers are accepted:
 * `12k` means 12Kb (12×1024)
 * `12m` means 12Mb (12×1024²)
 * `12g` means 12Gb (12×1024³)

The parameter argument can be given with or without space (`-b12m` is equivalent
to `-b 12m`).

`delay` is the desired delay, in milliseconds:
 * `12s` means 12 seconds (12×1000)
 * `12m` means 12 minutes (12×60×1000)
 * `12h` means 12 hours (12×60×60×1000)

The maximum expected bitrate provided by `delay` is the buffersize divided by
the delay.  For instance, `delay -b 10m 5s` will provide a maximum bitrate of
2Mb/s.

### Example

To delay the output of `command_A` to `command_B` by 5 seconds:

    command_A | delay 5s | command_B

If we need a 10Mb buffer:

    command_A | delay -b10m 5s | command_B

As an illustration, the following command produces an input at several rates,
and prints the result, delayed by 1 second, tabulated:

    { for i in {1..15}; do sleep $(bc <<< "scale=1;$i/10"); echo $i; done } |
        tee /dev/stderr | sed -u 's/^/\t/' | delay 1s

The following command allows a live-delayed video stream: it captures the webcam
and encodes the result in mpeg2, which is played in `vlc` with a delay of 2
seconds:

    ffmpeg -an -s 320x240 -f video4linux2 -i /dev/video0 -f mpeg2video -b 1M - |
        delay 2s | vlc -

## Blog post

<http://blog.rom1v.com/2014/01/lecture-differee-de-la-webcam-dun-raspberry-pi/>
(in French)


debian/changelog
delay (1.0-1) unstable; urgency=medium

  * Initial release

 -- Romain Vimont <rom@rom1v.com>  Fri, 06 Oct 2017 14:36:36 +0200


dtbuf.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "dtbuf.h"

#define DTBUF_CHUNK_PAYLOAD_SIZE 4000

struct header {
    time_ms timestamp;
    chunk_length data_length;
};

#define DTBUF_CHUNK_SIZE (sizeof (struct header) + DTBUF_CHUNK_PAYLOAD_SIZE)

int dtbuf_init(struct dtbuf *dtbuf, size_t capacity)
{
    // to avoid splitting a chunk on the circular buffer boundaries, add
    // (DTBUF_CHUNK_SIZE-1) bytes at the end: a chunk starting at (capacity-1)
    // will still fit
    dtbuf->real_capacity = capacity + DTBUF_CHUNK_SIZE - 1;
    if (!(dtbuf->data = malloc(dtbuf->real_capacity))) {
        return 1;
    }
    dtbuf->capacity = capacity;
    dtbuf->head = 0;
    dtbuf->tail = 0;
    return 0;
}

void dtbuf_free(struct dtbuf *dtbuf)
{
    free(dtbuf->data);
}

int dtbuf_is_empty(struct dtbuf *dtbuf)
{
    return dtbuf->head == dtbuf->tail;
}

int dtbuf_is_full(struct dtbuf *dtbuf)
{
    // When dtbuf->head >= dtbuf->capacity, it "cycles" (reset to 0) if and
    // only if there is enough space at the start for a full chunk.
    // Thus, if dtbuf->head has not cycled while it is after capacity, then the
    // buffer is full.
    // Else, if head >= tail, there is always enough space (by design).
    // Else (if head < tail), there is enough space only if dtbuf->tail is far
    // enough (ie we can put a full chunk at the start).
    return dtbuf->head >= dtbuf->capacity || (dtbuf->head < dtbuf->tail
                                              && dtbuf->tail - dtbuf->head <=
                                              DTBUF_CHUNK_SIZE);
}

time_ms dtbuf_next_timestamp(struct dtbuf *dtbuf)
{
    struct header *header = (struct header *) &dtbuf->data[dtbuf->tail];
    return header->timestamp;
}

ssize_t dtbuf_write_chunk(struct dtbuf *dtbuf, int fd_in, time_ms timestamp)
{
    ssize_t r;
    struct header header;
    // directly write to dtbuf, at the right index
    int payload_index = dtbuf->head + sizeof(struct header);
    if ((r = read(fd_in, &dtbuf->data[payload_index],
                  DTBUF_CHUNK_PAYLOAD_SIZE)) > 0) {
        // write headers
        header.timestamp = timestamp;
        header.data_length = (chunk_length) r;
        memcpy(&dtbuf->data[dtbuf->head], &header, sizeof(header));
        dtbuf->head = payload_index + r;
        if (dtbuf->head >= dtbuf->capacity && dtbuf->tail >= DTBUF_CHUNK_SIZE) {
            // not enough space at the end of the buffer, cycle if there is
            // enough at the start
            dtbuf->head = 0;
        }
    } else if (r == -1) {
        perror("read()");
    }
    return r;
}

ssize_t dtbuf_read_chunk(struct dtbuf *dtbuf, int fd_out)
{
    ssize_t w;
    struct header *pheader = (struct header *) &dtbuf->data[dtbuf->tail];
    struct header header;
    chunk_length length = pheader->data_length;
    // directly read from dtbuf, at the right index
    int payload_index = dtbuf->tail + sizeof(struct header);
    if ((w = write(fd_out, &dtbuf->data[payload_index], length)) > 0) {
        if (w == length) {
            // we succeed to write all the data
            dtbuf->tail = payload_index + w;
            if (dtbuf->tail >= dtbuf->capacity) {
                // the next chunk cannot be after capacity
                dtbuf->tail = 0;
                if (dtbuf->head >= dtbuf->capacity) {
                    // can happen if capacity < DTBUF_CHUNK_SIZE
                    dtbuf->head = 0;
                }
            }
        } else {
            dtbuf->tail += w;
            // set the timestamp for writing at the new tail position
            header.timestamp = pheader->timestamp;
            // set the remaining length
            header.data_length = length - w;
            memcpy(&dtbuf->data[dtbuf->tail], &header, sizeof(header));
        }
        if (dtbuf->head >= dtbuf->capacity && dtbuf->tail >= DTBUF_CHUNK_SIZE) {
            // there is enough space at the start now, head can cycle
            dtbuf->head = 0;
        }
    } else if (w == -1) {
        perror("write()");
    }
    return w;
}


dtbuf.h
#ifndef __H_DTBUF
#define __H_DTBUF

#include <sys/types.h>
#include <stddef.h>
#include <stdint.h>
#include "time_ms.h"

/*
 * dtbuf means either "direct timestamped buffer" or "delta-t buffer".
 * - direct: it directly reads from and writes to a FILE
 * - timestamped: every chunk is preceded by a header containing the date of
 *                read
 * - delta-t: the purpose of this buffer is to keep a constant delay between
 *            reads from stdin and writes to stdout
 *
 * In order to avoid chunks to be split at the circular buffer boundaries, its
 * real capacity is larger than the declared one, so that every read/write of
 * DTBUF_CHUNK_SIZE can be done in once, without splitting.
 */

typedef uint16_t chunk_length;

struct dtbuf {
    char *data;
    size_t capacity;            // expected capacity
    size_t real_capacity;       // capacity + DTBUF_CHUNK_SIZE - 1
    int head;                   // index of the next chunk
    int tail;                   // index of the oldest chunk in memory
};

// init a dtbuf structure with an expected capacity
int dtbuf_init(struct dtbuf *dtbuf, size_t capacity);

// free dtbuf content (not dtbuf itself)
void dtbuf_free(struct dtbuf *dtbuf);

// read the timestamp from the headers of the next chunk to be read
// assumes !is_empty(dtbuf)
time_ms dtbuf_next_timestamp(struct dtbuf *dtbuf);

// indicates whether the buffer is empty
int dtbuf_is_empty(struct dtbuf *dtbuf);

// indicates whether there is not enough space for writing a new full chunk
int dtbuf_is_full(struct dtbuf *dtbuf);

// read a chunk from fd_in and write it to dtbuf
// assumes fd_in is ready and !is_full(dtbuf)
ssize_t dtbuf_write_chunk(struct dtbuf *dtbuf, int fd_in, time_ms timestamp);

// read a chunk from dtbuf and write it to fd_out
// must be called when the next chunk time is reached
// assumes fd_out is ready and !is_empty(dtbuf)
ssize_t dtbuf_read_chunk(struct dtbuf *dtbuf, int fd_out);

#endif


main.c
#include <poll.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "dtbuf.h"
#include "errno.h"
#include "time_ms.h"

/*
 * Write to stdout DELAY ms later the data received on stdin.
 */

struct dtbuf dtbuf;

// stdout is at index 0 because we will always poll it
struct pollfd fds_all[] = {
    {1 /* stdout */ , POLLOUT},
    {0 /* stdin */ , POLLIN}
};

// default values, changed by parse_cli()
int delay = 5000;               // in ms
size_t dtbufsize = 1024 * 1024; // in bytes

void print_syntax(char *arg0)
{
    fprintf(stderr, "Syntax: %s [-b <dtbufsize>] <delay>\n", arg0);
}

// parse command: delay [-b <dtbufsize>] <delay_ms>
// these commands should work:
//   - delay 5s
//   - delay -b 10m 4000
//   - delay 4k -b 10000k
void parse_cli(int argc, char *argv[])
{
#define SYNTAX_ERROR 9
#define OPT_NONE 0
#define OPT_DTBUFSIZE 1
#define ARG_VALUE 1
#define ARG_DTBUFSIZE (1 << 1)
    int ctx = OPT_NONE;
    char *arg;
    int args_handled = 0;
    char *garbage;
    int i;
    for (i = 1; i < argc; i++) {
        arg = argv[i];
 handle_arg:
        if (ctx == OPT_DTBUFSIZE) {
            if (args_handled & ARG_DTBUFSIZE) {
                // dtbufsize parameter present twice
                print_syntax(argv[0]);
                exit(SYNTAX_ERROR);
            }
            // parse dtbufsize parameter
            errno = 0;
            dtbufsize = strtol(arg, &garbage, 0);
            if (errno) {
                perror("strtol()");
                exit(10);
            }
            // handle modifiers for Kb, Mb and Gb
            if (strcmp("k", garbage) == 0) {
                dtbufsize <<= 10;
            } else if (strcmp("m", garbage) == 0) {
                dtbufsize <<= 20;
            } else if (strcmp("g", garbage) == 0) {
                dtbufsize <<= 30;
            } else if (garbage[0] != '\0') {
                fprintf(stderr, "dtbufsize value contains garbage: %s\n",
                        garbage);
                exit(SYNTAX_ERROR);
            }
            args_handled |= ARG_DTBUFSIZE;
            ctx = OPT_NONE;
        } else if (strncmp("-b", arg, 2) == 0) {
            ctx = OPT_DTBUFSIZE;
            if (strlen(&arg[2]) > 0) {
                // handle "-b12" like "-b 12"
                arg = &arg[2];
                // yeah, this is ugly, but straightforward!
                goto handle_arg;
            }
        } else {
            if (args_handled & ARG_VALUE) {
                // delay value present twice
                print_syntax(argv[0]);
                exit(SYNTAX_ERROR);
            }
            // parse delay value
            errno = 0;
            delay = strtol(arg, &garbage, 0);
            if (errno) {
                perror("strtol()");
                exit(10);
            }
            // handle modifiers for seconds, minutes and hours
            if (strcmp("s", garbage) == 0) {
                delay *= 1000;
            } else if (strcmp("m", garbage) == 0) {
                delay *= 60 * 1000;
            } else if (strcmp("h", garbage) == 0) {
                delay *= 60 * 1000 * 1000;      // likely to be useless
            } else if (garbage[0] != '\0') {
                fprintf(stderr, "delay value contains garbage: %s\n", garbage);
                exit(SYNTAX_ERROR);
            }
            args_handled |= ARG_VALUE;
        }
    }
}

int main(int argc, char *argv[])
{
    int r;                      // returned by poll()
    int fds_count;
    int has_next_chunk = 0;
    time_ms next_chunk_timestamp = 0;
    time_ms now;
    time_ms wait_delay;
    // Note that stdin and stdout polling is not symetrical: we always poll
    // stdout for detecting POLLERR or POLLHUP (in that case, we want to stop
    // immediately), and we set POLLOUT only when we need to write.
    // On the contrary, we poll stdin only when we need to read (we don't want
    // to stop if we have chunks to write, even if stdin is closed).
    int poll_stdin;
    int pollout_stdout;
    int timeout;
    int in_closed = 0;
    int out_closed = 0;

    parse_cli(argc, argv);

    if (dtbuf_init(&dtbuf, dtbufsize)) {
        fprintf(stderr, "dtbuf initialization failed\n");
        exit(1);
    }

    do {
        // current time
        now = get_time_ms();

        // we want to poll stdin when we can store what we will read
        poll_stdin = !in_closed && !dtbuf_is_full(&dtbuf);

        // here, out_closed is always false
        // we want to pollout stdout if we have a next chunk to write now
        // if we only have a next chunk to write later, we set a timeout
        // instead
        if (has_next_chunk) {
            wait_delay = next_chunk_timestamp + delay - now;
            if (wait_delay <= 0) {
                // data to write as soon as possible
                pollout_stdout = 1;
                timeout = -1;
            } else {
                // data to write later
                pollout_stdout = 0;
                timeout = wait_delay;
            }
        } else {
            // no data to write at all
            pollout_stdout = 0;
            timeout = -1;
        }

        // we always want to poll stdout for detecting POLLERR or POLLHUP
        fds_count = poll_stdin ? 2 : 1;
        fds_all[0].events = pollout_stdout ? POLLOUT : 0;

        // poll() the selected fds
        if ((r = poll(fds_all, fds_count, timeout)) == -1) {
            perror("poll()");
            exit(3);
        }

        if (r == 0) {
            // timeout occurs: we need to write the next chunk to stdout
            // read from dtbuf and write to stdout
            if (dtbuf_read_chunk(&dtbuf, 1 /* stdout */ ) <= 0) {
                out_closed = 1;
            }
        } else {
            if (fds_all[0].revents) {
                // stdout has revents
                if (fds_all[0].revents & (POLLOUT | POLLHUP)) {
                    // read from dtbuf and write to stdout
                    if (dtbuf_read_chunk(&dtbuf, 1 /* stdout */ ) <= 0) {
                        out_closed = 1;
                    }
                } else {
                    out_closed = 1;
                }
            }
            if (poll_stdin && fds_all[1].revents) {
                // stdin has revents
                if (fds_all[1].revents & (POLLIN | POLLHUP)) {
                    // we may have waited, get the new current time
                    now = get_time_ms();
                    // read from stdin and write to dtbuf
                    if (dtbuf_write_chunk(&dtbuf, 0 /* stdin */ , now) <= 0) {
                        in_closed = 1;
                    }
                } else {
                    in_closed = 1;
                }
            }
        }

        // update next_chunk state
        has_next_chunk = !dtbuf_is_empty(&dtbuf);
        if (has_next_chunk) {
            next_chunk_timestamp = dtbuf_next_timestamp(&dtbuf);
        }
        // always stop on out_closed
        // also stop when there will be no more chunks anymore
    } while (!out_closed && (!in_closed || has_next_chunk));

    dtbuf_free(&dtbuf);
    return 0;
}
